copy/multiple.go

```diff
@@ -27,23 +31,115 @@ const (
 type instanceCopy struct {
 	op           instanceCopyKind
 	sourceDigest digest.Digest
+
+	// Fields which can be used by callers when operation
+	// is `instanceCopyClone`
+	cloneCompressionVariant OptionCompressionVariant
+	clonePlatform           *imgspecv1.Platform
+	cloneAnnotations        map[string]string
+}
+
+// internal type only to make imgspecv1.Platform comparable
+type platformComparable struct {
+	architecture string
+	os           string
+	osVersion    string
+	osFeatures   string
+	variant      string
+}
+
+// Converts imgspecv1.Platform to a comparable format.
+func platformV1ToPlatformComparable(platform imgspecv1.Platform) platformComparable {
+	osFeatures := slices.Clone(platform.OSFeatures)
+	sort.Strings(osFeatures)
+	return platformComparable{architecture: platform.Architecture,
+		os:         platform.OS,
+		osFeatures: strings.Join(osFeatures, ","),
+		osVersion:  platform.OSVersion,
+		variant:    platform.Variant,
+	}
+}
+
+// platformCompressionMap prepares a mapping of platformComparable -> CompressionAlgorithmNames for given digests
+func platformCompressionMap(list internalManifest.List, instanceDigests []digest.Digest) (map[platformComparable]*set.Set[string], error) {
+	res := make(map[platformComparable]*set.Set[string])
+	for _, instanceDigest := range instanceDigests {
+		instanceDetails, err := list.Instance(instanceDigest)
+		if err != nil {
+			return nil, fmt.Errorf("getting details for instance %s: %w", instanceDigest, err)
+		}
+		platform := platformV1ToPlatformComparable(*instanceDetails.ReadOnly.Platform)
+		platformSet, ok := res[platform]
+		if !ok {
+			platformSet = set.New[string]()
+			res[platform] = platformSet
+		}
+		platformSet.AddSlice(instanceDetails.ReadOnly.CompressionAlgorithmNames)
+	}
+	return res, nil
+}
+
+func validateCompressionVariantExists(input []OptionCompressionVariant) error {
+	for _, option := range input {
+		if option.Algorithm != nil {
+			_, err := compression.AlgorithmByName(option.Algorithm.Name())
+			if err != nil {
+				return fmt.Errorf("invalid algorithm %s in option.EnsureCompressionVariantsExist: %w", option.Algorithm.Name(), err)
+			}
+		}
+	}
+	return nil
 }
```

Comment:
Yes this is only to validate invalid `Algorithm` values,  as requested in comment here: https://github.com/containers/image/pull/1987#discussion_r1222041486

Reply:
That was necessary when the value was a string and we were checking only for `"zstd"` (and we would silently do nothing if the user asked for `"my-private-new-compression"`).

Now, I think we can… more or less… assume that `Algorithm` values come from `AlgorithmByName`, and if the user asked for an unsupported algorithm, that function would fail.

So that error condition went away by restructuring the code, and I don’t think an explicit check is necessary. But it also doesn’t hurt anything…